File: golang/selection-sort/selection-sort.go
package bubble_sort

func findSmallest(arr []int) int {
	smallest := arr[0]
	smallestIndex := 0
	for i := 1; i < len(arr); i++ {
		if arr[i] < smallest {
			smallest = arr[i]
			smallestIndex = i
		}
	}
	return smallestIndex
}

func SelectionSort(list []int) []int {
	var ordenedList []int = make([]int, 0, len(list)) // Criar slice com capacidade, mas sem elementos inicialmente

	for len(list) > 0 {
		smallest := findSmallest(list)
		ordenedList = append(ordenedList, list[smallest])
		list = append(list[:smallest], list[smallest+1:]...)
	}
	return ordenedList
}

File: typescript/src/@selection-sort/selection-sort.ts
export class SelectionSort {
  public sort(arr: Array<number>): Array<number> {
    const newArr = new Array<number>();
    const length = arr.length;

    for (let i = 0; i < length; i++) {
      const smallest = this.findSmallest(arr);
      newArr.push(arr.splice(smallest, 1)[0]);
    }
    return newArr;
  }

  private findSmallest(arr: Array<number>): number {
    let smallest = arr[0];
    let smallestIndex = 0;

    for (let i = 0; i < arr.length; i++) {
      if (arr[i] < smallest) {
        smallest = arr[i];
        smallestIndex = i;
      }
    }
    return smallestIndex;
  }
}
